File: BoundednessAnalysis/func_findNDShifting.m
function [m, info] = func_findNDShifting(model,option)
% solve trapping region condition using SDP. 
%
% Corresponding to Section 3.1 of Liao et. al, 2024

    if nargin < 2
        option.round_Ndigit = 3;
        option.verbose = false;
        option.tol = 1e-6;
    end
    
    % initialize output
    m = nan;
    info.feasibility = false;
    
    %% system param
    nx = model.nx;
    Ls = model.Ls;
    Q = model.Q;
    
    Inx = eye(nx);
    
    %% filtering the data matrix
    % if don't filter, set option.round_Ndigit = nan.
    if ~isnan(option.round_Ndigit)
        Ls = round(Ls, option.round_Ndigit);
        Q = round(Q, option.round_Ndigit);
    end
    
    %% solve by SDP
    % This SDP push As most further to negative definite (the most positive
    % eigenvalue is minimized). 
    cvx_begin sdp quiet
    cvx_solver mosek
%     cvx_solver SeDuMi
        variable a(1,1);
        variable m(nx,1);
        dual variable W;

        minimize(a);

        subject to 
            As = Ls;
            for i = 1:nx
                As = As - m(i)*Q(:,:,i);
            end
            W : As <= Inx*a;
    cvx_end
    
    % The SDP can be unbounded, i.e., a*=-inf. 
    % Redo the SDP with regularization of the coordinate shifting 
    if strcmp(cvx_status, 'Unbounded')
        warning('Unregularized TRSDP is unbounded below(a*=-inf). Rerun TRSDP with regularization');
        cvx_begin sdp quiet
        cvx_solver mosek
    %     cvx_solver SeDuMi
            variable m_bounded(nx,1);
            dual variable W;

            minimize(norm(m_bounded));

            subject to 
                As = Ls;
                for i = 1:nx
                    As = As - m_bounded(i)*Q(:,:,i);
                end
                W : As <= -1e-3*Inx;
        cvx_end
        
        m = m_bounded;
        a = max(eig(As));
    end
    
    % setting output
    if strcmp(cvx_status, 'Solved')        
        info.feasibility = true;
        info.a = a;
        info.As = As;
        info.W = W;
        
        if a < 0
            info.existTR = true;
            if option.verbose
                disp('Coordinated shift s.t. As<0 is found:');
                disp(m')
                fprintf('Most positive eigenvalue a = %.4f\n', a)
            end
        else
            info.existTR = false;
        end
    else
        if option.verbose
            disp(['cvx_status: ', cvx_status]);
            disp('could not find m such that As<0');
        end
    end
    
    info.m = m;
end

File: BoundednessAnalysis/func_findNDShifting.py
import numpy as np
import cvxpy as cp
from typing import Dict, Tuple, Optional

def func_findNDShifting(model: Dict, option: Optional[Dict] = None) -> Tuple[np.ndarray, Dict]:
    """
    Solve trapping region condition using SDP.
    Corresponding to Section 3.1 of Liao et. al, 2024
    """
    # Default options if not provided
    if option is None:
        option = {
            'round_Ndigit': 3,
            'verbose': False,
            'tol': 1e-6
        }
    
    # Initialize output
    m = np.nan
    info = {'feasibility': False}
    
    # System parameters
    nx = model.nx
    Ls = model.Ls
    Q = model.Q
    
    Inx = np.eye(nx)
    
    # Filtering the data matrix
    if not np.isnan(option['round_Ndigit']):
        Ls = np.round(Ls, option['round_Ndigit'])
        Q = np.round(Q, option['round_Ndigit'])
    
    # Solve by SDP
    # This SDP pushes As most further to negative definite
    a = cp.Variable(1)
    m_var = cp.Variable((nx, 1))
    
    As = Ls
    for i in range(nx):
        As = As - m_var[i] * Q[:,:,i]
    
    constraints = [As <= a * Inx]
    prob = cp.Problem(cp.Minimize(a), constraints)
    
    # Try solving with MOSEK
    try:
        prob.solve(solver=cp.MOSEK, verbose=False)
    except:
        # Fallback to default solver if MOSEK not available
        prob.solve(verbose=False)
    
    # Handle unbounded case
    if prob.status == 'unbounded':
        print('Warning: Unregularized TRSDP is unbounded below(a*=-inf). Rerun TRSDP with regularization')
        
        m_bounded = cp.Variable((nx, 1))
        As = Ls
        for i in range(nx):
            As = As - m_bounded[i] * Q[:,:,i]
        
        constraints = [As <= -1e-3 * Inx]
        prob = cp.Problem(cp.Minimize(cp.norm(m_bounded)), constraints)
        
        try:
            prob.solve(solver=cp.MOSEK, verbose=False)
        except:
            prob.solve(verbose=False)
            
        m = m_bounded.value
        a_val = np.max(np.linalg.eigvals(As.value))
    else:
        m = m_var.value
        a_val = a.value[0] if hasattr(a.value, '__len__') else a.value
    
    # Setting output
    if prob.status == 'optimal':
        info['feasibility'] = True
        info['a'] = a_val
        info['As'] = As.value
        info['W'] = constraints[0].dual_value
        
        if a_val < 0:
            info['existTR'] = True
            if option['verbose']:
                print('Coordinated shift s.t. As<0 is found:')
                print(m.T)
                print(f'Most positive eigenvalue a = {a_val:.4f}')
        else:
            info['existTR'] = False
    else:
        if option['verbose']:
            print(f'cvx_status: {prob.status}')
            print('could not find m such that As<0')
    
    info['m'] = m
    return m, info 